DOC-278 vertx (#1623)
hazelcast/hz-docs: docs/modules/integrate/pages/integrate-with-vertx.adoc

Vert.x is a reactive application toolkit for creating resource-efficient, concurrent, asynchronous and flexible applications on the JVM.

Hazelcast integrates with Vert.x in the form of a cluster manager — the link:https://vertx.io/docs/vertx-hazelcast/java/[Hazelcast Cluster Manager].

In Vert.x, a cluster manager is used for various functions including:

- Discovery and group membership of Vert.x nodes in a cluster.
- Maintaining cluster-wide topic subscriber lists (so we know which nodes are interested in which event bus addresses).
- Distributed Map support.
- Distributed Locks.
- Distributed Counters.

There are 2 modules to choose from:

> - `io.vertx:vertx-hazelcast` — this module is part of Vert.x and is maintained by the Vert.x team with contributions from Hazelcast developers. This module is licensed under the Apache 2 license.

Reviewer: Should we be stating the license of a third party software? That _could_ change in future.

Author: JamesG advises we leave this in for now

- `com.hazelcast:vertx-hazelcast-enterprise` — this module is built on top of `vertx-hazelcast` and leverages functionality of {enterprise-product-name} to implement some of the cluster manager functionality (e.g. it uses the CP Subsystem to implement strongly consistent Lock and Counter data structures: `io.vertx.core.shareddata.Lock` and `io.vertx.core.shareddata.Counter`).

> TIP: To learn how to add the vertx-hazelcast module and enable distributed session management, as well as how to implement a unique id generator, see xref:get-started-with-vertx.adoc[Get started with Vert.x].

Reviewer 2: ```suggestion
TIP: To learn how to add the vertx-hazelcast module and enable distributed session management, as well as how to implement a unique ID generator, see xref:get-started-with-vertx.adoc[Get started with Vert.x].
```

=== Versioning and Hazelcast compatibility

The Vert.x Hazelcast Enterprise module follows the versioning of Vert.x. If you use an `x.y.z` version of Vert.x, you should use an `x.y.z` version of `vertx-hazelcast-enteprise`.

The `vertx-hazelcast-enteprise` module is compatible with all Hazelcast versions supported at the time of the release of the `vertx-hazelcast-enteprise` module, unless stated otherwise.

While older versions may work, such configurations are not supported.

== Use Vert.x Hazelcast Cluster Manager

For reference documentation for the `vertx-hazelcast` module, see the Vert.x Hazelcast Cluster Manager site.

You can also follow our xref:get-started-with-vertx.adoc[Get started with Vert.x guide].

== Use a different Hazelcast version

Due to Java compatibility reasons, the Vert.x Hazelcast module does not depend on the latest version of Hazelcast.

> You can change the Hazelcast dependency to any version of Hazelcast you need, e.g. you can change it to Hazelcast 5.2.x for Java 8 compatibility, or to the latest.

Reviewer: Should we be suggesting an out-of-support version of Hazelcast, or keep it abstract?

Author: JamesG advises we remove the "e.g. you can change it to Hazelcast 5.2.x for Java 8 compatibility, or to the latest" since 5.3 will be out of support in 7 days. This will be part of a future discussion though

> NOTE: Older versions may not be supported by Hazelcast and do not receive any patches.

Reviewer 2: ```suggestion
NOTE: Older versions may not be supported by Hazelcast or receive any patches.
```
I don't think we can state definitively that any older version doesn't receive patches
